lib/plugins/rotate_asg_instances/asg.rb

```diff
@@ -0,0 +1,237 @@
+require_relative 'system'
+
+module Moonshot
+  class ASG # rubocop:disable Metrics/ClassLength
+    include Moonshot::CredsHelper
+
+    def initialize(resources)
+      @resources = resources
+      @ilog = @resources.ilog
+      @ssh_user = ENV['MOONSHOT_SSH_USER'] || ENV['LOGNAME']
+    end
+
+    def asg
+      @asg ||= Aws::AutoScaling::AutoScalingGroup.new(name: name)
+    end
+
+    def rotate_asg_instances
+      @ilog.start_threaded('Rotating ASG instances...') do |step|
+        @step = step
+        @volumes_to_delete = outdated_volumes(outdated_instances)
+        @shutdown_instances = cycle_instances(outdated_instances)
+        @step.success('ASG instances rotated successfully!')
+      end
+    end
+
+    def teardown_outdated_instances
+      @ilog.start_threaded('Tearing down outdated instances...') do |step|
+        @step = step
+        terminate_instances(@shutdown_instances)
+        reap_volumes(@volumes_to_delete)
+        @step.success('Outdated instances removed successfully!')
+      end
+    end
+
+    def name
+      @resources.controller.stack
+                .resources_of_type('AWS::AutoScaling::AutoScalingGroup')
+                .first.physical_resource_id
+    end
+
+    def outdated_instances
+      asg.instances.reject do |i|
+        i.launch_configuration_name == asg.launch_configuration_name
+      end
+    end
+
+    private
+
+    def outdated_volumes(instances = outdated_instances)
```

Comment:
Re raising the comment that we can skip the argument here and directly using outdated_instances inside the method.

Reply:
It means updating the outdated_instances method too.
```ruby
        def outdated_instances
          @outdated_instances ||= asg.instances.reject do |i|
            i.launch_configuration_name == asg.launch_configuration_name
         end
       end
```

Reply:
Made the changes suggested here.